design/sync-labels-bmh-to-node.md

```diff
@@ -0,0 +1,374 @@
+2. For each BMH, fetch corresponding Node using the mapping discussed above.
+3. Synchronize the labels and update the Node in the workload cluster.
+
+An alternative would be to set up a watch on the Node resource types in every
+workload cluster. The benefit being that we can quickly synchronize any label
+changes on the Node object itself. However, this introduces complications when
+the management cluster != workload cluster and additionally, when there are
+many workload clusters being managed by a single management cluster. How such
+a watch could be implemented within the controller-runtime model would require
+further investigation.
+
+### Configuration
+
+#### Command-Line Flag
+
+We need to expose two additional configuration variables to the user: the set
+of label prefixes and the full-sync interval. We can introduce two new
+command-line flags for CAPM3, and also utilize default values if the flags are
+not specified: (i) `--bmh-label-prefixes` with a nil default value and (ii)
+`--label-sync-interval` with a 15-second default value.
+
+Unfortunately, command-line flags will not be compatible with openShift where
+cluster administrators do not deploy metal3 themselves and do not have access
+to the command line arguments for the cluster-api provider.
+
+#### Annotations on BMH
+
+In this approach, an annotation is added to a BMH, in addition to the labels
+to be synchronized, which reflects the prefixes to be utilized. For example,
+
+```yaml
+  kind: BareMetalHost
+  metadata:
+    annotations:
+      metal3.io/label-prefixes: {"my-prefix.metal3.io",
+      "my-prefix.kubernetes.io"}
+    labels:
+      my-prefix.metal3.io/rack: xyz-123
+      my-prefix.kubernetes.io/zone: security-level-0
+```
+
+A concern with this approach is consistency since the same annotation would
+have to be duplicated across BMHs. The prefixes, while user configurable, are
+static and should not require re-configuring. For example, in a likely
+scenario, the cluster admin would define a standard set of prefixes to
+be used internally. Also, annotations can be (accidentally) added/removed/
+changed, leading to undesirable behavior.
+
+#### New API Type
+
+We could introduce a new API type: `BareMetalHostLabelSyncProfile` (any other
+suggestions for the name?).
+
+The common case is when a label on the BMH is updated, we must first find the
+corresponding `BareMetalHostLabelSyncProfile` (if any) that matches the BMH.
+To avoid fetching all `BareMetalHostLabelSyncProfile`, we can make use of a
+label on the BMH, in namespace/name format, that references the
+`BareMetalHostLabelSyncProfile` (see example below). We can drop the namespace
+requirement if it's in the same namespace as the BMH.
+
+The less common case is where the `BareMetalHostLabelSyncProfile` is updated (
+e.g. a prefix is removed). In this case, we will need to fetch all BMHs
+associated and perform a sync operations across the impacted Nodes. Label
```

Comment:
What would the sync do? Remove the previously synced labels from the nodes?

Reply:
AFIK, when a prefix is removed, it's not entirely clear which prefix was removed (i.e. we can't do a diff against earlier state of the object). 
This make removal of labels associated with the removed prefix difficult if not impossible. Any existing labels will be left as is (the user can remove them). I documented this in the "Prefix Deletion" section below. 

When a prefix is added, any existing labels matching the prefix will be synced immediately.

Reply:
That doesn't seem ideal. Using a separate CRD to configure the controller to sync 1 prefix would let us trap the delete and handle it.

Reply:
If I understand correctly, you are suggesting creating a new API type where each object of that type will be a specific prefix. 

Couple of comments:
1. While this would allow us to catch the above delete case, this would require more work to put together. My thought was to simply document the procedure. I would image this being a rare situation -- I can't image a scenario where users need to frequently update prefix values. 
2. More importantly, how do you envision mapping a "prefix object" to the set of BMHs it should apply to? With Metal3Cluster we get this logical mapping for free: Metal3Cluster -> Metal3Machine/Machine -> BMH. I suppose we could use labels, much like we do with the other API type I described in the doc; however, the Metal3Cluster grouping seems more appropriate to me as it does not require the user/admin to explicitly add labels to resources to keep them linked.

Reply:
The new resource could either link to a Metal3Cluster or use a labelSelector to choose hosts. It may be more work, initially, but would provide a more robust solution. We can start with the annotation approach, and add an API later, I suppose.

Reply:
Sounds good. The initial approach utilizes the annotation but I'm happy to follow up with an API approach if there is desire for it.